packages/next/src/server/use-cache/use-cache-wrapper.ts

```diff
@@ -695,8 +699,16 @@ export function cache(kind: string, id: string, fn: any) {
 
             await ignoredStream.cancel()
           }
-
-          stream = entry.value
+          // If we have a cache scope, we need to clone the entry and set it on
+          // the inner cache scope.
+          else if (resumeDataCache) {
+            const split = cloneCacheEntry(entry)
+            stream = split[0].value
+            resumeDataCache.cache.set(
+              serializedCacheKey,
+              Promise.resolve(split[1])
```

Comment:
I still don't quite think this is safe or at least I can't convince myself it is.

Currently we have an in-memory LRU for this cache and so any currently generating stream will have to buffer before resolving the cache entry promise in the first place because that's how it gets the tags/revalidate time etc.

However, if this was a remove cache handler. For example, imagine it doing a fetch() for the cache entry and getting the tags/revalidate etc from `response.headers` but the stream from `response.body`. Then we can resolve this promise with a cache entry before having loaded the whole stream.

We would've done all this in the prospective render.

Before the stream finishes reading we could've started the prerender which would now try to read this stream even though it's not done which would treat it as dynamic.

I think we need to wait to call

```
cacheSignal.endRead()
```

Until we've buffered the full stream in this case.